backend/build.gradle

```diff
@@ -41,17 +38,41 @@ dependencies {
     runtimeOnly 'com.mysql:mysql-connector-j'
     annotationProcessor 'org.projectlombok:lombok'
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
-    testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc'
     testRuntimeOnly 'org.junit.platform:junit-platform-launcher'
+
+    // rest assured
     testImplementation 'io.rest-assured:rest-assured:5.5.0'
+
+    // Spring REST Docs
+    asciidoctorExt 'org.springframework.restdocs:spring-restdocs-asciidoctor:3.0.1'
+    testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc:3.0.1'
+
+}
+
+tasks.named('jar') {
+    enabled = false
 }
 
 tasks.named('test') {
-    outputs.dir snippetsDir
     useJUnitPlatform()
+    reports.html.required = false
+    reports.junitXml.required = false
+    outputs.dir snippetsDir
 }
 
 tasks.named('asciidoctor') {
-    inputs.dir snippetsDir
     dependsOn test
+    inputs.dir snippetsDir
+    configurations 'asciidoctorExt'
+    sources {
+        include("**/index.adoc")
+    }
+    baseDirFollowsSourceFile()
+}
+
+tasks.named('bootJar') {
+    dependsOn asciidoctor
+    from("${asciidoctor.outputDir}") {
+        into 'static/docs'
+    }
```

Comment:
의견) 지금처럼 static/docs에 두게되면 매핑해주는 컨트롤러가 필요할 것 같은데, 아래처럼 바꿔서 index 페이지를 문서로 사용하게 하는 건 어떨까요??

```gradle
tasks.named('asciidoctor') {
    dependsOn test
    inputs.dir snippetsDir
    configurations 'asciidoctorExt'
    sources {
        include("**/index.adoc")
    }
    baseDirFollowsSourceFile()
}

task createDocument(type: Copy) {
    dependsOn asciidoctor

    from file("build/docs/asciidoc/index.html")
    into file("src/main/resources/static")
}

tasks.named('bootJar') {
    dependsOn createDocument
}
```

Reply:
static resource 경로이기 떄문에 컨트롤러가 따로 필요없는 것으로 알고 있어요!
그래서 로컬에서 jar파일을 실행하면 `localhost:8080/docs/index.html`로 바로 접근 가능합니당

```gradle
task createDocument(type: Copy) {
    dependsOn asciidoctor

    from file("build/docs/asciidoc/index.html")
    into file("src/main/resources/static")
}
```
이든이 작성해주신 코드는
```gradle
tasks.named('bootJar') {
    dependsOn asciidoctor
    from("${asciidoctor.outputDir}") {
        into 'static/docs'
    }
}
```
이 코드와 같은 기능을 수행하기 때문에 이대로 가도 괜찮지 않을까 생각해요!

Reply:
<img width="767" alt="스크린샷 2024-07-27 23 42 31" src="https://github.com/user-attachments/assets/365adca6-4d91-4af9-93a1-0ac59c97bd8c">

프린은 접속 잘 되시나요??

Reply:
<img width="2224" alt="스크린샷 2024-07-27 오후 11 48 16" src="https://github.com/user-attachments/assets/6a399f39-6b3d-4ae2-86b5-fcff85be3c92">

옙

Reply:
저는 왜 안될까요 ㅠㅠ

Reply:
제 브랜치에서 bootJar를 실행하셨을까요?

Reply:
넴넴 그렇게 했습니다

Reply:
아 원인 발견했습니다.
어떤 이유인지는 모르겠지만 bootJar 돌려도 static 밑에 docs/index.html 이 생성되지 않고 있네요
<img width="329" alt="스크린샷 2024-07-27 23 51 47" src="https://github.com/user-attachments/assets/d598e739-d8c8-444a-8873-c9a6c1605747">

Reply:
```yml
# db, jpa 설정
cors:
  origin: "*"

logging:
  level:
    web: TRACE
    org:
      hibernate:
        orm:
          jdbc:
            bind: TRACE
      springframework:
        transaction:
          interceptor: TRACE
```
이렇게 application-local.yml 설정 후

```sh
./gradlew clean bootJar

java -jar ./build/libs/ddangkong-0.0.1-SNAPSHOT.jar 
```
실행해서 로그 보여주실 수 있나요?

Reply:
> 아 원인 발견했습니다. 어떤 이유인지는 모르겠지만 bootJar 돌려도 static 밑에 docs/index.html 이 생성되지 않고 있네요 <img alt="스크린샷 2024-07-27 23 51 47" width="329" src="https://private-user-images.githubusercontent.com/84304802/352759936-d598e739-d8c8-444a-8873-c9a6c1605747.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIwOTIyNDMsIm5iZiI6MTcyMjA5MTk0MywicGF0aCI6Ii84NDMwNDgwMi8zNTI3NTk5MzYtZDU5OGU3MzktZDhjOC00NDRhLTg4NzMtYzlhNmMxNjA1NzQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI3VDE0NTIyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQyZDcyMDA1OWU5NDI0NGVhNDk0MGE4YWNkY2I5OTM5MTIzNzU1NWI1ZjM3ZjIxOTg3YTM0NzcxZmIwZDRmZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6J02rm4cxDKyMsn6LPL7O5vMMOrSu9ARsSgdh3eTVjE">

booJar task에서 `static/docs`로 copy하는 것은 jar 파일에만 존재하기 때문에 아마 이든의 ide에는 보이지 않을거에요

Reply:
넵 잠시만용

Reply:
```
jun@Junui-MacBookPro backend % ./gradlew clean bootJar

java -jar ./build/libs/ddangkong-0.0.1-SNAPSHOT.jar

> Task :compileJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
OpenJDK 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended
<=======<========-----> 63% E<========-----> 63% EXEC<========----2024-07-27T23:54:41.584+09:00  INFO 75549 --- [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'est > 2> :test > 25 > :test >> :test > 37 tests completed
2024-07-27T23:54:41.584+09:00  INFO 75549 --- [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Shutdown initiated...
2024-07-27T23:54:41.586+09:00  INFO 75549 --- [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Shutdown completed.투표_생성
2024-07-27T23:54:41.591+09:00  INFO 75549 --- [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2024-07-27T23:54:41.591+09:00  INFO 75549 --- [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : HikariPool-2 - Shutdown initiated...
2024-07-27T23:54:41.592+09:00  INFO 75549 --- [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : HikariPool-2 - Shutdown completed.
2024-07-27T23:54:41.596+09:00  INFO 75549 --- [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2024-07-27T23:54:41.596+09:00  INFO 75549 --- [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2024-07-27T23:54:41.597+09:00  INFO 75549 --- [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : HikariPool-3 - Shutdown initiated...
2024-07-27T23:54:41.597+09:00  INFO 75549 --- [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : HikariPool-3 - Shutdown completed.

> Task :asciidoctor
2024-07-27T23:54:42.336+09:00 [main] WARN FilenoUtil : Native subprocess control requires open access to the JDK IO subsystem
Pass '--add-opens java.base/sun.nio.ch=ALL-UNNAMED --add-opens java.base/java.io=ALL-UNNAMED' to enable.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.8/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 9s
9 actionable tasks: 9 executed

```

Reply:
deprecated gradle feature가 사용되고 있다네요..? 관련있을까요

Reply:
지금은 무시해도 좋을 것 같아요!

```sh
java -jar ./build/libs/ddangkong-0.0.1-SNAPSHOT.jar 
```
명령어 실행해서 localhost:8080/docs/index.html 로 접근해볼까용??

Reply:
똑같이 접근은 안됩니다ㅠ
<img width="769" alt="image" src="https://github.com/user-attachments/assets/20fb15a3-7d2e-4187-8771-cbf896a59735">

Reply:
jar 실행하면 콘솔에 요청 로그 찍힐텐데 어떤 로그인지 볼 수 있을까요??

Reply:
2024-07-27T23:59:01.846+09:00 DEBUG 75554 --- [nio-8080-exec-5] .m.m.a.ExceptionHandlerExceptionResolver : Resolved [org.springframework.web.servlet.NoHandlerFoundException: No endpoint GET /docs/index.html.]

Reply:
요건데요 요거때문에 매핑해줘야하는거 아닌가 여쭤봤어요

Reply:
디코 음성채널 와주실 수 있나요?

Reply:
작업실로 오시죠

Reply:
✨✨ 디코 화면공유로 해결했읍니다

> 아래처럼 바꿔서 index 페이지를 문서로 사용하게 하는 건 어떨까요??
> ```gradle
> task createDocument(type: Copy) {
>     dependsOn asciidoctor
>
>    from file("build/docs/asciidoc/index.html")
>    into file("src/main/resources/static")
> }
>```

현재 개발 서버는 public subnet의 ec2 인스턴스에 존재하며, 프론트엔드의 요청뿐 아니라 불특정 다수의 부적절한 접근 시도도 함께 받게 됩니다(해킹을 위한 ip 무차별 대입 공격 등)

만일 루트 경로(`https://ddangkong.io/index.html`)에 API 문서를 노출할 경우 해당 경로에 쉽게 접근할 수 있어 다른 API까지 노출될 위험이 있습니다

docs 경로를 추가(`https://ddangkong.io/docs/index.html`)하는 것이 완벽한 방어책이 될 순 없으나, 루트 경로보다 상대적으로 엔드포인트를 추측하기 어렵게 때문에 지금의 엔드포인트를 유지하기로 결정합니다